core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationDecider.java

```diff
@@ -115,7 +115,7 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
             case NONE:
                 return allocation.decision(Decision.NO, NAME, "no allocations are allowed");
             case NEW_PRIMARIES:
-                if (shardRouting.primary() && shardRouting.allocatedPostIndexCreate(indexMetaData) == false) {
+                if (shardRouting.primary() && shardRouting.active() == false && shardRouting.recoverySource().initialRecovery()) {
```

Comment:
I'm OK with the semantics change but we should document this, if you haven't done it already.